packages/mobx-react/__tests__/observer.test.tsx

```diff
@@ -908,3 +908,72 @@ test("Missing render should throw", () => {
     }
     expect(() => observer(Component)).toThrow()
 })
+
+test("this.context is observable if ComponentName.contextType is set", () => {
+    const Context = createContext({})
+
+    let renderCounter = 0
+
+    @observer
```

Comment:
Thanks this test looks a lot better! Still want to point out that this component doesn't have to be an observer at all if you'd put an observable object as context value, saving re-rendering the root of the tree and any unmemo-ed child thereunder, whenever you increase the counter. That is the way it is typically done, and I suspect also the reason why no one reported this earlier.

Reply:
What difference does it make to the render passes whether the context value is in an observable or in a computed?

The component must react either way to pass the new value to the context.

Reply:
In your setup, it is `Parent` that reacts to a change in `counter`, by creating a new object (at line 929) as part of the render process to compute `this.contextValue` (you basically make an unnecessary state copy here). If you'd pass a stable observable object (or even `this`) as Provider value, the counter itself isn't read at all by the Parent, but instead by the Child, so that the effect of changing `counter` would be much more local effect. So instead of re-rendering in principle the whole subtree of Parent, you'd only be re-rendering individual Child components.

Reply:
So basically cutting the Reacts render logic completely out and using parent only as Context Provider? This is kind of strict.

Reply:
In real life I often have something like this:

```
  @computed
  get contextValue(): DropdownContextType {
    return {
      toggleDropdown: this.toggleDropdown,
      closeDropdown: this.closeDropdown,
      open: this.open,
      type: this.props.type || this.computedType,
    };
  }
  ```
  
 and `this.open` is also accessed in render to determine if dropdown should be rendered open or not. So it doesn't matter here.

Reply:
I don't see the benefit of constructing that object over and over again over just passing `this` (you strongly typed out the other members anyway). If you really want to strict down the object passed, you could also allocate it once and just make it a delegator object, so that you can keep using the same `contextValue` even when you're state changes over time:

```
contextValue = {
  toggleDropdown: this.toggleDropdown,
  closeDropdown: this.closeDropdown,
  get open() { return this.open }, 
  get type() { return this.props.type || this.computedType }
}
```

> So basically cutting the Reacts render logic completely out and using parent only as Context Provider?

If there is no need for it yes, if there is need for it, fine as well. In you're cases shown so far there is no need (as in, there is no visual changes to your UI 'above' the Child components.

Reply:
But this way you get into nasty `this` bugs.

https://codesandbox.io/s/admiring-stallman-hcnrpv?file=/src/Form.tsx

Reply:
Thank you for your help and support! And for Mobx obviously! Luckily it came out early enough so that I didn't dipped to much in Redux.